### PR TITLE
Add support for variadic functions in the catalog and builtin CONCAT function

### PR DESCRIPTION
**Pull request by jkosh44**

### CODE REVIEW

File: src/catalog/database_catalog.cpp
Comment (mbutrovich) [inline src/catalog/database_catalog.cpp:2205]: I'm surprised we do so much registering of abort actions to cleanup pointers when other parts of catalog bootstrapping asserts that we can't fail. Do we envision this bootstrap txn ever failing, @tanujnay112?

**Review by pmenon — CHANGES_REQUESTED** [review]
So, CONCAT is actually a variable argument function. Is this something the catalog supports? If so, can you implement that version? The TPL API becomes: `@concat(str1, str2, str2, ...)` and the C++ API is something like `StringFunctions::Concat(StringVal *result, ExecutionContext *ctx, StringVal **inputs, uint32_t num_inputs)`. It should be pretty simple to implement.

**Comment by jkosh44** [comment]
> So, CONCAT is actually a variable argument function. Is this something the catalog supports? If so, can you implement that version? The TPL API becomes: `@concat(str1, str2, str2, ...)` and the C++ API is something like `StringFunctions::Concat(StringVal *result, ExecutionContext *ctx, StringVal **inputs, uint32_t num_inputs)`. It should be pretty simple to implement.

I'll take a look at this and see

File: src/catalog/database_catalog.cpp
Comment (jkosh44) [reply src/catalog/database_catalog.cpp:2205]: > I'm surprised we do so much registering of abort actions to cleanup pointers when other parts of catalog bootstrapping asserts that we can't fail. Do we envision this bootstrap txn ever failing, @tanujnay112?

I'm not really sure why all the abort actions are there. I tested it and it does work fine without them. Would the DB still even be able to start up if this txn aborted?

**Comment by jkosh44** [comment]
> > So, CONCAT is actually a variable argument function. Is this something the catalog supports? If so, can you implement that version? The TPL API becomes: `@concat(str1, str2, str2, ...)` and the C++ API is something like `StringFunctions::Concat(StringVal *result, ExecutionContext *ctx, StringVal **inputs, uint32_t num_inputs)`. It should be pretty simple to implement.
> 
> I'll take a look at this and see

@pmenon So I went to look in the catalog and it doesn't look like we support functions with variable argument lenght
https://github.com/cmu-db/terrier/blob/dabbbef03678288cd4a6b4a34284055dbaa3f0c1/src/include/catalog/database_catalog.h#L309-L311
https://github.com/cmu-db/terrier/blob/dabbbef03678288cd4a6b4a34284055dbaa3f0c1/src/include/catalog/database_catalog.h#L333-L335

I think this is probably blocked until that functionality gets added. It doesn't seem like something trivial to add either.

**Comment by jkosh44** [comment]
I'm closing this for now to avoid Jenkins builds

File: src/catalog/postgres/builder.cpp
Comment (jkosh44) [inline src/catalog/postgres/builder.cpp:769]: I'm honestly not super confident on these boolean parameters, if someone could double check them.

File: src/catalog/database_catalog.cpp
Comment (jkosh44) [inline src/catalog/database_catalog.cpp:3005]: Is this accurate, or do we need to also check that all arguments have the same type?

File: src/catalog/database_catalog.cpp
Comment (tanujnay112) [reply src/catalog/database_catalog.cpp:2207]: At that moment, I just put this in because every other Bootstrapping function was doing so as well. But yeah now that I think about it we shouldn't have the bootstrapping transaction fail. Otherwise, the whole system pretty much does not proceed from there right.

File: src/include/execution/vm/bytecode_handlers.h
Comment (tanujnay112) [inline src/include/execution/vm/bytecode_handlers.h:1493]: Other functions in this file that pass in arrays and such don't pass in the length of the array or in this case number of inputs as a pointer but the actual value. Is there a specific reason this is a pointer by any chance?

File: src/include/execution/sql/functions/string_functions.h
Comment (tanujnay112) [inline src/include/execution/sql/functions/string_functions.h:33]: Don't use references on primitive types.

File: src/execution/vm/bytecode_generator.cpp
Comment (tanujnay112) [inline src/execution/vm/bytecode_generator.cpp:3507]: It seems like you're using this for the right-hand side of an assignment. Use VisitExpressionForRValue instead as it looks like this being treated like an R-value.

**Review by tanujnay112 — CHANGES_REQUESTED** [review]

File: src/catalog/database_catalog.cpp
Comment (jkosh44) [reply src/catalog/database_catalog.cpp:2207]: If that's the case I can remove this line from the PR, and I can open a separate PR to remove it from the other functions

File: src/include/execution/vm/bytecode_handlers.h
Comment (jkosh44) [reply src/include/execution/vm/bytecode_handlers.h:1493]: No good reason, I'll de-reference it in vm.cpp so it's not a pointer here.

File: src/execution/vm/bytecode_generator.cpp
Comment (jkosh44) [reply src/execution/vm/bytecode_generator.cpp:3507]: So for this I'm a little confused tbh. @pmenon can you weigh in here? I thought that these were supposed to be LValues and we were mistakenly using RValues. @pmenon recently opened this issue: https://github.com/cmu-db/noisepage/issues/1167

File: src/execution/sema/sema_builtin.cpp
Comment (pmenon) [inline src/execution/sema/sema_builtin.cpp:2600]: Fix index for error message.

File: src/execution/sql/functions/string_functions.cpp
Comment (pmenon) [inline src/execution/sql/functions/string_functions.cpp:13]: The `inputs` array needs to store *pointers* to string values to avoid copying each string value into the array. So:

```
Concat(StringVal *result, ExecutionContext *ctx, const StringVal *inputs[], uint32_t num_inputs)
```

File: src/execution/sql/functions/string_functions.cpp
Comment (pmenon) [inline src/execution/sql/functions/string_functions.cpp:14]: Technically, this shouldn't be possible. It will be caught earlier during planning or even in code-gen. I'd change this to an assert ensuring at least one input.

File: src/execution/sql/functions/string_functions.cpp
Comment (pmenon) [inline src/execution/sql/functions/string_functions.cpp:21]: You need to only count (and copy) non-NULL inputs.

File: src/execution/sql/functions/string_functions.cpp
Comment (pmenon) [inline src/execution/sql/functions/string_functions.cpp:32]: I don't think `CONCAT` ever returns NULL.

File: src/execution/vm/bytecode_generator.cpp
Comment (pmenon) [inline src/execution/vm/bytecode_generator.cpp:2357]: The length is a compile-time constant, directly emit the value in the bytecode rather than assigning to a temporary. Also, using an 8-byte count may be overkill?

File: src/execution/vm/bytecode_generator.cpp
Comment (pmenon) [reply src/execution/vm/bytecode_generator.cpp:3507]: SQL values should use L-values. But, this function shouldn't make assumptions on how to derive child expressions. Either (1) add a boolean indicating whether to use L- or R-values, or (2) pull this logic into `CONCAT` generation. I'd prefer the latter until we have enough use-cases for this to make refactor into a common function.

File: src/execution/vm/vm.cpp
Comment (pmenon) [inline src/execution/vm/vm.cpp:2117]: Change to (1) use an immediate for the input count, and (2) use a proper array of pointers for the inputs.

File: src/include/execution/sql/functions/string_functions.h
Comment (pmenon) [inline src/include/execution/sql/functions/string_functions.h:33]: Ditto, change signature.

File: src/include/type/type_id.h
Comment (pmenon) [inline src/include/type/type_id.h:34]: This isn't a real `Type`, but neither is `PARAMETER_OFFSET`. Will need to refactor later.

**Review by pmenon — CHANGES_REQUESTED** [review]
I think you'll need to add additional type-checking on these variadics to ensure someone doesn't pass non-strings into `CONCAT`.

If it's a significant amount of work, I'd prefer creating an issue and punt for later.

File: src/include/type/type_id.h
Comment (jkosh44) [reply src/include/type/type_id.h:34]: So the actual type from postgres is "any" (https://github.com/cmu-db/noisepage/issues/1176). We can just rename this to that?

File: src/execution/vm/bytecode_generator.cpp
Comment (jkosh44) [reply src/execution/vm/bytecode_generator.cpp:2357]: When you say it's a compile-time constant you're talking about when compiling the tpl, not when compiling the C++ right? How would I emit the number directly in the bytecode?

**Comment by jkosh44** [comment]
> I think you'll need to add additional type-checking on these variadics to ensure someone doesn't pass non-strings into `CONCAT`.
> 
> If it's a significant amount of work, I'd prefer creating an issue and punt for later.

I'm not really sure what you mean by this to be honest. I thought that `sema_builtin` was doing exactly this?

File: src/include/execution/sql/functions/string_functions.h
Comment (jkosh44) [reply src/include/execution/sql/functions/string_functions.h]: Updated

File: src/include/execution/vm/bytecode_handlers.h
Comment (jkosh44) [reply src/include/execution/vm/bytecode_handlers.h]: Updated

File: src/execution/sema/sema_builtin.cpp
Comment (jkosh44) [reply src/execution/sema/sema_builtin.cpp]: Updated

File: src/execution/sql/functions/string_functions.cpp
Comment (jkosh44) [reply src/execution/sql/functions/string_functions.cpp]: updated

File: src/execution/sql/functions/string_functions.cpp
Comment (jkosh44) [reply src/execution/sql/functions/string_functions.cpp]: Updated

File: src/execution/sql/functions/string_functions.cpp
Comment (jkosh44) [reply src/execution/sql/functions/string_functions.cpp]: Updated

File: src/execution/sql/functions/string_functions.cpp
Comment (jkosh44) [reply src/execution/sql/functions/string_functions.cpp]: Updated

File: src/execution/vm/bytecode_generator.cpp
Comment (jkosh44) [reply src/execution/vm/bytecode_generator.cpp]: I pulled it into CONCAT generation, and kept it as L-Value (which I think is what you were saying).

File: src/execution/vm/vm.cpp
Comment (jkosh44) [reply src/execution/vm/vm.cpp]: Updated (2), waiting on input for (1) from other question.

File: src/include/execution/sql/functions/string_functions.h
Comment (jkosh44) [reply src/include/execution/sql/functions/string_functions.h]: updated

File: cmake_modules/FindLibev.cmake
Comment (pmenon) [inline cmake_modules/FindLibev.cmake:1]: Accidental include?

File: src/execution/sql/functions/string_functions.cpp
Comment (pmenon) [inline src/execution/sql/functions/string_functions.cpp:24]: Technically, `offset` should be a `std::size_t`. I'd also go ahead and change the type on line 17 for cosistency.

File: src/execution/vm/bytecode_generator.cpp
Comment (pmenon) [reply src/execution/vm/bytecode_generator.cpp:2357]: > When you say it's a compile-time constant you're talking about when compiling the tpl, not when compiling the C++ right?

Correct.

> How would I emit the number directly in the bytecode?

Take a look at how `EmitDerefN()` works. You should create a new function `BytecodeEmitter::EmitConcat()` that takes in the argument count as a `uint32_t`.

File: src/include/execution/sql/functions/string_functions.h
Comment (pmenon) [inline src/include/execution/sql/functions/string_functions.h:33]: Should make the input count unsigned, probably a `uint32_t`.

File: src/include/type/type_id.h
Comment (pmenon) [reply src/include/type/type_id.h:34]: I'd say leave as it. This'll need to be refactored later outside this PR.

**Review by pmenon — CHANGES_REQUESTED** [review]
Small changes, then it's probably good to go.

File: cmake_modules/FindLibev.cmake
Comment (jkosh44) [reply cmake_modules/FindLibev.cmake:1]: oops yeah

Comment (jkosh44) [reply cmake_modules/FindLibev.cmake]: removed

File: src/execution/sql/functions/string_functions.cpp
Comment (jkosh44) [reply src/execution/sql/functions/string_functions.cpp]: Fixed

File: src/execution/vm/bytecode_generator.cpp
Comment (jkosh44) [reply src/execution/vm/bytecode_generator.cpp]: Updated

File: src/include/execution/sql/functions/string_functions.h
Comment (jkosh44) [reply src/include/execution/sql/functions/string_functions.h]: Fixed

**Comment by jkosh44** [comment]
@pmenon I believe I've addressed all your comments except for 

> I think you'll need to add additional type-checking on these variadics to ensure someone doesn't pass non-strings into `CONCAT`.
> 
> If it's a significant amount of work, I'd prefer creating an issue and punt for later.

I'm still not really sure what this entails.

File: src/execution/vm/vm.cpp
Comment (pmenon) [inline src/execution/vm/vm.cpp:2117]: Nit: try to be consistent with parameter ordering between bytecode and C++ functions.

**Review by pmenon — CHANGES_REQUESTED** [review]
You also need to modify the `Concat` bytecode encoding in `bytecodes.h` [here](https://github.com/cmu-db/noisepage/blob/master/src/include/execution/vm/bytecodes.h#L595) so that `num_inputs` is an immediate unsigned 4-byte integer (e.g., like the last operand to `CSVReaderInit` [here](https://github.com/cmu-db/noisepage/blob/master/src/include/execution/vm/bytecodes.h#L493))

**Comment by pmenon** [comment]
Don't worry about variadic function type-check for now. Ideally, this checking should be done during function lookup in binding. Having a check during `TPL` semantic analysis is good for non-SQL use-cases (i.e., UDFs), but moving as much checking earlier in the pipeline is generally the correct approach.

File: src/execution/vm/vm.cpp
Comment (jkosh44) [reply src/execution/vm/vm.cpp]: Updated

**Comment by jkosh44** [comment]
> You also need to modify the `CONCAT` bytecode encoding in `bytecodes.h` [here](https://github.com/cmu-db/noisepage/blob/master/src/include/execution/vm/bytecodes.h#L595) so that `num_inputs` is an immediate unsigned 4-byte integer (e.g., like the last operand to `CSVReaderInit [here](https://github.com/cmu-db/noisepage/blob/master/src/include/execution/vm/bytecodes.h#L493))

Updated

**Review by pmenon — APPROVED** [review]
LGTM

File: src/catalog/postgres/builder.cpp
Comment (mbutrovich) [reply src/catalog/postgres/builder.cpp:769]: Those look reasonable.

**Comment by mbutrovich** [comment]
> Don't worry about variadic function type-check for now. Ideally, this checking should be done during function lookup in binding. Having a check during `TPL` semantic analysis is good for non-SQL use-cases (i.e., UDFs), but moving as much checking earlier in the pipeline is generally the correct approach.

Let's open an issue for this once this gets merged.

**Review by lmwnshn — APPROVED** [review]
Just echoing the LGTM.

**Pull request merged**
